src/System.Console/src/System/Console.cs

```diff
@@ -85,6 +89,55 @@ public static void ResetColor()
             ConsolePal.ResetColor();
         }
 
+        // This is the worker delegate that is called on the Threadpool thread to fire the actual events. It must guarantee that it
+        // signals the caller on the ControlC thread so that it does not block indefinitely.
+        private static void ControlCDelegate(object data)
+        {
+            ControlCDelegateData controlCData = (ControlCDelegateData)data;
+            try
+            {
+                controlCData.DelegateStarted = true;
+                ConsoleCancelEventArgs args = new ConsoleCancelEventArgs(controlCData.ControlKey);
+                controlCData.CancelCallbacks(null, args);
+                controlCData.Cancel = args.Cancel;
+            }
+            finally
+            {
+                controlCData.CompletionEvent.Set();
+            }
+        }
+
+        public static event ConsoleCancelEventHandler CancelKeyPress
+        {
+            add
+            {
+                lock (InternalSyncObject)
+                {
+                    _cancelCallbacks += value;
+
+                    // If we haven't registered our control-C handler, do it.
+                    if (_registrar == null)
+                    {
+                        _registrar = new ConsolePal.ControlCHandlerRegistrar();
+                        _registrar.Register();
+                    }
+                }
+            }
+            remove
+            {
+                lock (InternalSyncObject)
+                {
+                    // If count was 0, call SetConsoleCtrlEvent to remove cb.
+                    _cancelCallbacks -= value;
+                    Debug.Assert(_cancelCallbacks == null || _cancelCallbacks.GetInvocationList().Length > 0, "Teach Console::CancelKeyPress to handle a non-null but empty list of callbacks");
+                    if (_registrar != null && _cancelCallbacks == null)
+                    {
+                        _registrar.Unregister();
```

Comment:
Does _registrar need to be nulled out here?  Otherwise it seems like the next time a handler is registered with CancelKeyPress, _cancelCallbacks will be updated from null to non-null, but since _registrar will still be non-null, Register won't be called.